onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -191,21 +192,52 @@ CUDAExecutionProvider::PerThreadContext::~PerThreadContext() {
 }
 
 bool CUDAExecutionProvider::PerThreadContext::IsGraphCaptureAllowed() const {
-  return regular_run_count_before_graph_capture_ >= min_num_runs_before_cuda_graph_capture_;
+  return regular_run_count_before_graph_capture_ >= min_num_runs_before_cuda_graph_capture_ &&
+         IsGraphCaptureAllowedOnRun();
+}
+
+bool CUDAExecutionProvider::PerThreadContext::IsGraphCaptureAllowedOnRun() const {
+  return cuda_graph_.IsGraphCaptureAllowedOnRun();
+}
+
+Status CUDAExecutionProvider::PerThreadContext::SetCudaGraphAnnotationId(
+    const onnxruntime::RunOptions& run_options) {
+  auto graph_annotation_str =
+      run_options.GetConfigOptions().GetConfigEntry(kOrtRunOptionsConfigCudaGraphAnnotation);
+  // If graph annotation is not provided, fall back to the one cuda graph per session behavior
+  CudaGraphAnnotation_t cuda_graph_annotation_id = 0;
+  if (graph_annotation_str.has_value()) {
+    ORT_TRY {
+      cuda_graph_annotation_id = std::stoi(graph_annotation_str.value());
+    }
+    ORT_CATCH(const std::invalid_argument&) {
```

Comment:
nit: parse integer options is very common, may add a function in ConfigOptions so that everyone can call it.

Reply:
using functions in ConfigOptions from cuda_execution_provider requires exposing through provider bridges. It seems try_stoi() is not generic enough to be done this way.